### PR TITLE
Use Repository pattern to decouple persistence from domain logic

### PR DESCRIPTION
**Pull request by johngallagher**
* We're using the repository pattern as part of the dashboard work
* This was an experiment before introducing it to the rest of the team
* Now let's talk about how/if we could use this in other places

### CODE REVIEW

File: _patterns/use-repository-pattern-to-decouple.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+## The problem
+
+The ActiveRecord pattern that Rails uses mixes two different concerns:
+
+* Domain logic
+* Persistence
+
+## The symptoms
+
+* Large God classes - `SocialUser`
```
Comment (bockets) [inline _patterns/use-repository-pattern-to-decouple.md:15]:
```suggestion
* Large classes
```

**Comment by johngallagher** [comment]
> Thank you for putting this together John. I agree that there are problems with slow tests, but it's not a huge concern. And I do not think that ActiveRecord is to blame for that.

Right now, you're right. When you look in CI the tests that are taking the most time have nothing to do with ActiveRecord - they're end to end tests that are inefficient and test things that don't need to be tested / should be tested at a different level.

The main benefit of this pattern isn't speed. It's modularity of design.

> 
> As it stands, I think that the problems that you're trying to solve are too great for one "patten" document and I would prefer to discuss those problems when they arise rather than by trying to introduce a Pattern.

OK - so where would you suggest a document around the benefits of the repository pattern should live?

We already have a bunch of these problems with the billing area of the codebase - what's the correct place for having these discussions if not in the patterns and practices meeting?

**Comment by bockets** [comment]
> > Thank you for putting this together John. I agree that there are problems with slow tests, but it's not a huge concern.
> 
> With a test suite that takes 20 minutes to run over 31 parallel processes, I can't agree with that. This affects our cycle time (although definitely not as much as using / abusing PRs in the way we do for code review).

I want fast tests, everyone wants fast tests. That's the problem worth discussing, but not in our Patterns repo. One solution for that is "avoid IO in tests", and that could be fixed by adding a Repository pattern, but it can also be fixed in a ton of other ways, some that require much less code.

**Comment by johngallagher** [comment]
> > > Thank you for putting this together John. I agree that there are problems with slow tests, but it's not a huge concern.
> > 
> > 
> > With a test suite that takes 20 minutes to run over 31 parallel processes, I can't agree with that. This affects our cycle time (although definitely not as much as using / abusing PRs in the way we do for code review).
> 
> I want fast tests, everyone wants fast tests. That's the problem worth discussing, but not in our Patterns repo. One solution for that is "avoid IO in tests", and that could be fixed by adding a Repository pattern, but it can also be fixed in a ton of other ways, some that require much less code.

Yup. Fast tests isn't the main problem that repositories are trying to solve, so I've removed the references to slow tests.

They allow controlled upgrades to Stripe as per the example in this PR without making lots of changes across the codebase.

If you don't think repositories are a good solution to this, what would you do instead @bockets ?

**Comment by karlentwistle** [comment]
Given that we have recently introduced the repository pattern in our dashboard project, and considering that different teams will be contributing new widgets to the dashboard, it might be worthwhile to temporarily pause the rollout of the pattern to other areas of the codebase. This pause would provide an opportunity for each team to experience and evaluate the current implementation, allowing them to form their own opinions.

By giving each team a chance to work with the repository pattern, we can gather valuable feedback and insights from their firsthand experiences. This feedback will help us assess the pattern's suitability and effectiveness in different contexts, as well as identify any potential challenges or improvements that need to be addressed.

Taking this approach allows for a more informed decision-making process before fully integrating the repository pattern across the codebase. It promotes collaboration, empowers the teams to provide input, and ensures that any concerns or suggestions they may have are considered. Once each team has had the opportunity to utilize the current implementation, we can gather their feedback and collectively make an informed decision on the pattern's future usage in the broader codebase.

**Comment by bockets** [comment]
Thanks for the links @karlentwistle.

@johngallagher respectfully, what I see is searching for solutions to problems that we don't really have, or problems that do not really slow us down, except for a few really bad areas. This pattern really presents the Repository Pattern as a cure all. It's such a huge departure from where our application has gone for the last 13ish years that it would take a tremendous undertaking to go another direction. Meanwhile, the application would be a confusing mess of two patterns for a long time. Right now, I don't believe we have a problem that would be better solved with the Repository Pattern.

# Tight coupling

Except for a few places (Billing), I don't see a huge problem with tightly coupling our domain logic to our database. At the end of the day, logic has to live somewhere, and the least abstract place to put it is typically the ActiveRecord model. It may not be perfect, but it's what thousands of Rails developers have done and continue to do. The other option is just plain-ole Ruby objects.

# God classes

God classes happen. It's great when they don't, but at the same time, it's not really that big of a deal when they do. It's better to just be mindful and say "We should avoid adding more logic to classes that are already really large, and consider extracting logic if we can." But I don't think that the Repository Pattern helps us get there.

# Slow tests

I used to think that a +10-minute suite was a problem. Now, not so much. I haven't felt like it's slowed me down, and I close more PRs than anyone. I feel pretty confident in our tests. It would be great if people leaned more towards using `build` in our tests as the default rather than `create` so that we could avoid IO.

# Bounded contexts

I disagree that this is a problem we have in Ruby since everything is duck typed. If you want to set up barriers from parts of the application to be able to talk directly to other parts of the application, maybe you could justify doing that by passing configuration objects that wrap ActiveRecord models, such that the models could be swapped out easily with other objects.

# Conclusion

I want to encourage continuing to seeking out opportunities to learn and apply new patterns, but those proposals need to be carefully balanced against the additional complexity and onboarding time, especially when they deviate from what industry engineers are familiar with. I fear that my statements here may act to discourage exploration and discovery. Please don't feel like we shouldn't continue seeking out better ways to write code.

**Comment by johngallagher** [comment]
Thanks for your thoughtful and detailed comments @bockets and @DyegoCosta 

I'll close this avenue of exploration for now.

**Comment by johngallagher** [comment]
Also, I think we probably need to have an engineering wide discussion about consistency. Our styles vary wildly across different PRs and commits.

It might be that as a company we decide to fully embrace The Rails Way of doing things. It sounds like this might be the right thing to do for onboarding and training purposes.

**Pull request closed without merging**